presto-hive/src/main/java/com/facebook/presto/hive/S3SelectCsvRecordReader.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.CSVInput;
+import com.amazonaws.services.s3.model.CSVOutput;
+import com.amazonaws.services.s3.model.CompressionType;
+import com.amazonaws.services.s3.model.ExpressionType;
+import com.amazonaws.services.s3.model.InputSerialization;
+import com.amazonaws.services.s3.model.OutputSerialization;
+import com.amazonaws.services.s3.model.SelectObjectContentRequest;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.spi.type.TypeManager;
+import org.apache.hadoop.fs.Path;
+import parquet.hadoop.metadata.CompressionCodecName;
+
+import java.util.List;
+import java.util.Properties;
+
+import static org.apache.hadoop.hive.serde.serdeConstants.ESCAPE_CHAR;
+import static org.apache.hadoop.hive.serde.serdeConstants.QUOTE_CHAR;
+
+public class S3SelectCsvRecordReader
+        extends S3SelectLineRecordReader
+{
+    /*
+     * Sentinel unicode comment character (http://www.unicode.org/faq/private_use.html#nonchar_codes).
+     * It is expected that \uFDD0 sentinel comment character is not the first character in any row of user's CSV S3 object.
+     * The rows starting with \uFDD0 will be skipped by S3Select and will not be a part of the result set or aggregations.
+     * To process CSV objects that may contain \uFDD0 as first row character please disable S3Optimizer.
+     * TODO: Remove this proxy logic when S3Select API supports disabling of row level comments.
+     */
+
+    private static final String COMMENTS_CHAR_STR = "\uFDD0";
+
+    S3SelectCsvRecordReader(S3SelectClient selectClient, Path path, long start, long length,
+            Properties schema, List<HiveColumnHandle> columns,
+            TupleDomain<HiveColumnHandle> effectivePredicate, TypeManager typeManager)
+    {
+        super(selectClient, path, start, length, schema, columns, effectivePredicate, typeManager);
+    }
+
+    @Override
+    public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, String query, Path path)
+    {
+        String fieldDelimiter = getFieldDelimiter(schema);
+        String quoteChar = getQuoteChar(schema);
+        String escapeChar = getEscapeChar(schema);
+        SelectObjectContentRequest selectObjectRequest = new SelectObjectContentRequest();
+        AmazonS3URI uri = new AmazonS3URI(path.toUri());
```

Comment:
Seems like AmazonS3URI would fail with endpoints like "s3n://" which are currently supported.

Reply:
That's correct, used PrestoS3FileSystem helper methods to get bucket and key for s3n and s3a paths.